### DISCUSSION TITLE
Is embedding components dynamically possible?

### DISCUSSION THREAD

**Discussion by mpeterdev** [near/react-on-chain#101] [Implementation]
Near Social has a beta feature for embedding components in posts, here is the announcement message
![Screenshot 2023-10-18 at 9 45 07 AM](https://github.com/near/bos-web-engine/assets/24904709/900375bd-f82c-4077-bbfa-158a3201f2e2)
([post](https://near.social/mob.near/widget/MainPage.N.Post.Page?accountId=mob.near&blockHeight=103593826))

@andy-haynes would this be possible under the current architecture of BWE? If not, do you see blockers towards implementing it?

@thisisjoshford if this proves challenging, we need to identify the priority of supporting this. Here is a relevant section from Eugene's pinned post from last year on the future of Near Social
![Screenshot 2023-10-18 at 9 47 20 AM](https://github.com/near/bos-web-engine/assets/24904709/b803fd3c-7b01-4793-beeb-e2d1813e00b8)

**Comment by andy-haynes** [comment]
> would this be possible under the current architecture of BWE? If not, do you see blockers towards implementing it?

Looks feasible to me.

When rendering Components with data that potentially embeds other Components, we can parse them out during serialization. Since the Component would be dynamically loaded, they would always be sandboxed. Normally this would require a unique identifier, but since the embedding text is static we can generate those when parsing.

So given a Component that embeds content:
```jsx
<div>
  embedded [EMBED](ex.near/widget/Badge) Component!
</div>
```

This would currently be serialized as:
```js
{
  type: "div",
  props: {
    children: ["embedded [EMBED](ex.near/widget/Badge) Component!"],
  },
  componentChildren: [],
}
```

Which is not far off from what it would take to embed the Component currently:
```js
{
  type: "div",
  props: {
    children: [
      "embedded ",
      // the `-0` is appended to make the ID unique based on index within content
      { type: "div", props: { id: "dom-ex.near/widget/Badge-0" } },
      " Component!",
    ],
  },
  componentChildren: ["ex.near/widget/Badge"],
}
```

**Reply by mpeterdev** [reply]
just to clarify, this answer holds true even if the embed string is fetched dynamically from SocialDB e.g. as the content of a post, right?

I believe that is what you just stated when mentioning that it would always be dynamically loaded but want to be 100% clear

**Reply by andy-haynes** [reply]
>just to clarify, this answer holds true even if the embed string is fetched dynamically from SocialDB e.g. as the content of a post, right?

Yes, that's right.

>I believe that is what you just stated when mentioning that it would always be dynamically loaded but want to be 100% clear

It's definitely related - what I outlined would happen at render time when the element is serialized for sending via `postMessage`. Serialization is where `<Widget />` renders are transformed into placeholders in DOM for sandboxed Components, so this feature would complement the current approach pretty well.
